cares: sync with upstream, fully adopt v1.13.0

rvagg opened the pull request:
A reopen of #9332, I couldn't do it properly there so this is fresh. It also brings us up to 1.13.0 which includes the fix patched in #13897.

This is not complete unfortunately. I'm getting these:

```
=== release test-dns-cancel-reverse-lookup ===
Path: parallel/test-dns-cancel-reverse-lookup
assert.js:43
  throw new errors.AssertionError({
  ^

AssertionError [ERR_ASSERTION]: 'ENOTFOUND' === 'ECANCELLED'
    at QueryReqWrap.resolver.reverse.common.mustCall (/Users/rvagg/git/iojs/io.js/test/parallel/test-dns-cancel-reverse-lookup.js:14:12)
    at QueryReqWrap.callback (/Users/rvagg/git/iojs/io.js/test/common/index.js:509:15)
    at QueryReqWrap.onresolve [as oncomplete] (dns.js:246:10)
Command: out/Release/node /Users/rvagg/git/iojs/io.js/test/parallel/test-dns-cancel-reverse-lookup.js
```

and 

```
=== release test-dns-resolveany ===
Path: parallel/test-dns-resolveany
assert.js:684
assert.ifError = function ifError(err) { if (err) throw err; };
                                                  ^

Error: queryAny EBADRESP example.org
    at errnoException (dns.js:58:10)
    at QueryReqWrap.onresolve [as oncomplete] (dns.js:246:19)
Command: out/Release/node /Users/rvagg/git/iojs/io.js/test/parallel/test-dns-resolveany.js
```

I haven't been able to figure the first one out yet, nothing critical seems to have changed in that path as far as I can tell. I haven't looked at the second yet either.

addaleax commented:
I think you’re going to need to refloat https://github.com/c-ares/c-ares/commit/0ef4a0c64b6106a8102721e97c6b04b0f89bedd1 and https://github.com/c-ares/c-ares/commit/18ea99693d63f957ecb670045adbd2c1da8a4641 as well, that takes care of at least one of the issues

rvagg commented:
@addaleax aye! I just figured that out and have cherry picked them and am pushing for another CI. Looking good locally at least. Thanks.

rvagg commented:
OK, looking good @ https://ci.nodejs.org/job/node-test-commit/12334/, unrelated http2 failure on freebsd10 and unrelated fd failure on aix.

There's a bunch of things going on here and I probably need to do more squashing:

* Upgrade to c-ares 1.12.0, note that we've been running a really heavily patched version of c-ares for some time now, completely out of sync with upstream, just pulling in patches when we've needed or wanted them, so this gets us in properly in sync
* Upgrade to c-ares 1.13.0 on top of that
* Some automatic and some manual platform config files to match the new version(s)
* Updated license-builder.sh to use the new style license coming from upstream c-ares, and updated LICENSE from that too
* Applied an old floating patch on c-ares for a special Windows case that we haven't pushed upstream, ref #5090 (see https://github.com/nodejs/node/pull/5090/commits/fc1d66c8df6364e59ddc29257999135719a05c8c), this is pretty gross and I'm not sure it'd be accepted upstream or not. I think the commit needs renaming because we're going to be living with this one for a while.
* Applied the two new patches by @addaleax from #13883 and #15023, both of which are upstream but not part of 1.13.0 so I expect we'll only be living with these floating for a little while.
* Added a new patch that we may have to keep on floating if we don't find a better approach. In one of the recent versions, c-ares switched many natively named structs to their own custom forms, presumably for better cross-platform support. One of these is `ares_ssize_t` which is mostly an alias for `ssize_t`. The way we work with GYP means that this gets defined in the platform ares_config.h files in each of the platform directories inside of deps/cares/config/, _but_ they only get included when compiling c-ares itself and are not included when c-ares is included in src/ compilation. That's worked fine until now because a couple of these new definitions now live in ares_build.h and if you build c-ares stand-alone then you get the custom platform ares_config.h file included, but if you build it as an external library you don't get that platform config and that now matters. I've tried solving this a bunch of ways, including exposing the right ares_config.h via `direct_dependent_settings` in cares.gyp and then using ares_setup.h prior to loading ares.h in our env.h, and a bunch of other convoluted methods to get this right. In the end I've opted for simple manual `typedef`s, bypassing the default `#define` method, defaulting to `ssize_t` which is absolutely fine on all of our platforms except for Windows and there's a special case for it in there anyway.

PTAL

addaleax commented:
@bnoordhuis @indutny re: a2e37f48fe108f996143be66d41428d028f54622 … is this a patch we’re going to float for all of eternity? should we maybe rather try to upstream it, or work around it in Node’s code?

bnoordhuis commented:
@piscisaureus tried upstreaming it but it was never merged.  I remember it being rejected but maybe it simply fell by the wayside.  The mailing list thread is [here](https://c-ares.haxx.se/mail/c-ares-archive-2012-08/0003.shtml).

rvagg commented:
I've pushed a fresh version with maximal squashing, updated commit descriptions and summaries and a reimplementation of the older floating patch that works properly for 1.13.0, pretty close to original.
New CI to confirm https://ci.nodejs.org/job/node-test-commit/12349/
This is GTG as far as I'm concerned, would appreciate reviews.

Note that upgrading c-ares involves unpacking source, trimming out a ton of unnecessary files and manually sorting into src/ and include/, then updating the ares_config.h files—some of which are done on those platforms using `./configure` and some of which are done with a bit of guesswork (e.g. android). So reviews looking at the c-ares upgrade commits should look at the deps/cares/config/ changes in particular, they are the heaviest-touch parts of the upgrade, the rest is mostly just copying source and headers.

BridgeAR commented:
I guess this could land as is?

BridgeAR commented:
Ping @rvagg

jasnell commented:
Adding don't land labels for 8, 6 and 4 defensively. This could be landed in 8 if the @nodejs/lts WG agrees.

gibfahn commented:
Is this semver-minor?

addaleax commented:
@gibfahn We don’t ship the c-ares headers, so c-ares API changes aren’t changes of our API, so I’d go with “no”.

refack commented:
>> Is this semver-minor?

> so I’d go with “no”.

So `patch`?

BridgeAR commented:
If there is no disagreement from @nodejs/lts or @rvagg in the next 48 hours I would just go ahead and land this.

BridgeAR commented:
CI https://ci.nodejs.org/job/node-test-pull-request/10362/

BridgeAR commented:
Landed in 50e580df210109b87190d5dff84e0d62a74d60a9...a9f125449e61457cbbad79fbae766b603b861d1b

rvagg commented:
thanks @BridgeAR, I lost track of this one!

sgallagher commented:
@rvagg So what's the plan with regards to 8.x? Will this be landing there (since it's not public API)?

gibfahn commented:
@sgallagher do you want it in 8.x? As @jasnell said in https://github.com/nodejs/node/pull/15378#issuecomment-331947303, we can consider it if there's a reason for it.

sgallagher commented:
@gibfahn It would be a big help, as this would allow us in Fedora to switch back to using the system copy of c-ares rather than carrying it bundled in Node.js. We're carrying the bundle because some years ago I was told (I think by @bnoordhuis ?) that the bundled c-ares carried patches not in an upstream release, but if you're resyncing to the official 1.13 release, I know that we'll be able to link against the system copy safely.

sgallagher commented:
@gibfahn Sorry, I can't translate "this PR refloats a couple of patches so you might not be able to use it directly". Does that mean that it's going to patch c-ares to differ from upstream? Because if so, then it doesn't matter to me whether this gets backported because I'm going to be stuck bundling it anyway.

addaleax commented:
The patches that we don’t float upstream (af171b7 and 34d125f) look like they should affect behaviour/API on Windows only, so you’re good there.

For the other two … I would recommend waiting until they are in a c-ares release, yes.

bnoordhuis commented:
@gibfahn I think af171b7 can be obsoleted by a few tweaks to our build, which leaves 34d125f.  I offer to provide moral support for anyone willing to upstream that. :-)

<sup>(I'll do it if nobody else will.)</sup>
